### ISSUE TITLE
.NET Native Compilation causes a Database to read the wrong Property for a Property

### ISSUE THREAD

**Issue by WilliamABradley** [dotnet/efcore#8188]
Trying to fetch a string from an object in a DBContext causes a crash saying that it was receiving the wrong type. Currently my multiple DataBase Contexts are stored in a Separate Universal Windows Class Library from my Main App, as the Databases are also used inside the BackgroundTasks WinRTComponent.

Using Debug is fine, however, if I try with Release on x86 or x64 (Haven't tried ARM), I receive the below error when trying to access this specific function in the database, which is a static method inside the DBContext. The database is somewhat working (I see it partially populated with from what I can tell is the right data in the right fields), and the other Databases seem to work fine.

```
Exception message: 
An exception occured while reading a database value for property 'Series.Language'. The expected type was 'System.Nullable`1[TheTVDBSharp.Models.Language]' but the actual value was of type 'TheTVDBSharp.Models.ContentRating'.
```
```
Stack trace:
   at Microsoft.EntityFrameworkCore.Metadata.Internal.EntityMaterializerSource.ThrowReadValueException[TValue](Exception exception, Object value, IPropertyBase property)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.EntityMaterializerSource.TryReadValue[TValue](ValueBuffer valueBuffer, Int32 index, IPropertyBase property)
   at Internal.Runtime.Augments.RuntimeAugments.CallDescrWorker(IntPtr callDescr) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 855
   at Internal.Runtime.TypeLoader.CallConverterThunk.InvokeTarget(Void* allocatedStackBuffer, CallConversionParameters& conversionParams) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 925
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBufferInternal[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context, ConservativelyReportedRegionDesc& regionDesc) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 919
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBuffer[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 903
   at Internal.Runtime.TypeLoader.CallConverterThunk.CallConversionThunk(IntPtr callerTransitionBlockParam, IntPtr callConversionId) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 480
   at Internal.Runtime.Augments.RuntimeAugments.CallDescrWorker(IntPtr callDescr) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 855
   at Internal.Runtime.TypeLoader.CallConverterThunk.InvokeTarget(Void* allocatedStackBuffer, CallConversionParameters& conversionParams) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 925
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBufferInternal[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context, ConservativelyReportedRegionDesc& regionDesc) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 919
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBuffer[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 903
   at Internal.Runtime.TypeLoader.CallConverterThunk.CallConversionThunk(IntPtr callerTransitionBlockParam, IntPtr callConversionId) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 480
   at System.InvokeUtils.CallIHelperStaticCall(Object thisPtr, IntPtr methodToCall, IntPtr dynamicInvokeHelperMethod, ArgSetupState& argSetupState, Boolean isTargetThisCall)
   at System.InvokeUtils.CallDynamicInvokeMethod(Object thisPtr, IntPtr methodToCall, Object thisPtrDynamicInvokeMethod, IntPtr dynamicInvokeHelperMethod, IntPtr dynamicInvokeHelperGenericDictionary, Object defaultParametersContext, Object[] parameters, Boolean invokeMethodHelperIsThisCall, Boolean methodToCallIsThisCall) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\System\InvokeUtils.cs:line 314
   at System.Linq.Expressions.Interpreter.MethodInfoCallInstruction.InvokeWorker(Object[] args) in f:\dd\ndp\fxcore\src\System.Linq.Expressions\System\Linq\Expressions\Interpreter\CallInstruction.cs:line 380
   at System.Linq.Expressions.Interpreter.MethodInfoCallInstruction.Run(InterpretedFrame frame) in f:\dd\ndp\fxcore\src\System.Linq.Expressions\System\Linq\Expressions\Interpreter\CallInstruction.cs:line 420
   at System.Linq.Expressions.Interpreter.Interpreter.Run(InterpretedFrame frame) in f:\dd\ndp\fxcore\src\System.Linq.Expressions\System\Linq\Expressions\Interpreter\Interpreter.cs:line 106
   at System.Linq.Expressions.Interpreter.LightLambda.Run(Object[] arguments) in f:\dd\ndp\fxcore\src\System.Linq.Expressions\System\Linq\Expressions\Interpreter\LightLambda.cs:line 274
   at System.Func`2.InvokeObjectArrayThunk(T arg)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.UnbufferedEntityShaper`1.Shape(QueryContext queryContext, ValueBuffer valueBuffer)
   at Internal.Runtime.Augments.RuntimeAugments.CallDescrWorker(IntPtr callDescr) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 855
   at Internal.Runtime.TypeLoader.CallConverterThunk.InvokeTarget(Void* allocatedStackBuffer, CallConversionParameters& conversionParams) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 925
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBufferInternal[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context, ConservativelyReportedRegionDesc& regionDesc) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 919
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBuffer[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 903
   at Internal.Runtime.TypeLoader.CallConverterThunk.CallConversionThunk(IntPtr callerTransitionBlockParam, IntPtr callConversionId) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 480
   at Internal.Runtime.Augments.RuntimeAugments.CallDescrWorker(IntPtr callDescr) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 855
   at Internal.Runtime.TypeLoader.CallConverterThunk.InvokeTarget(Void* allocatedStackBuffer, CallConversionParameters& conversionParams) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 925
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBufferInternal[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context, ConservativelyReportedRegionDesc& regionDesc) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 919
   at Internal.Runtime.Augments.RuntimeAugments.RunFunctionWithConservativelyReportedBuffer[T](Int32 cbBuffer, IntPtr pfnTargetToInvoke, T& context) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\Internal\Runtime\Augments\RuntimeAugments.cs:line 903
   at Internal.Runtime.TypeLoader.CallConverterThunk.CallConversionThunk(IntPtr callerTransitionBlockParam, IntPtr callConversionId) in f:\dd\ndp\fxcore\src\System.Private.TypeLoader\Internal\Runtime\TypeLoader\CallConverterThunk.cs:line 480
   at Microsoft.EntityFrameworkCore.Query.QueryMethodProvider.<_ShapedQuery>d__3`1.MoveNext()
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
   at _$ILCT$.$ILT$ReflectionDynamicInvoke$.InvokeRetOI[T0,T1](Object thisPtr, IntPtr methodToCall, ArgSetupState& argSetupState, Boolean targetIsThisCall)
   at System.InvokeUtils.CallIHelperStaticCallWithInstantiation(Object thisPtr, IntPtr methodToCall, IntPtr dynamicInvokeHelperMethod, ArgSetupState& argSetupState, Boolean isTargetThisCall, IntPtr dynamicInvokeHelperGenericDictionary)
   at System.InvokeUtils.CallDynamicInvokeMethod(Object thisPtr, IntPtr methodToCall, Object thisPtrDynamicInvokeMethod, IntPtr dynamicInvokeHelperMethod, IntPtr dynamicInvokeHelperGenericDictionary, Object defaultParametersContext, Object[] parameters, Boolean invokeMethodHelperIsThisCall, Boolean methodToCallIsThisCall) in f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\System\InvokeUtils.cs:line 309
   at System.Linq.Expressions.Interpreter.MethodInfoCallInstruction.InvokeWorker(Object[] args) in f:\dd\ndp\fxcore\src\System.Linq.Expressions\System\Linq\Expressions\Interpreter\CallInstruction.cs:line 380
   at System.Linq.Expressions.Interpreter.MethodInfoCallInstruction.Run(InterpretedFrame frame) in f:\dd\ndp\fxcore\src\System.Linq.Expressions\System\Linq\Expressions\Interpreter\CallInstruction.cs:line 420
   at System.Linq.Expressions.Interpreter.Interpreter.Run(InterpretedFrame frame) in f:\dd\ndp\fxcore\src\System.Linq.Expressions\System\Linq\Expressions\Interpreter\Interpreter.cs:line 106
   at System.Linq.Expressions.Interpreter.LightLambda.Run(Object[] arguments) in f:\dd\ndp\fxcore\src\System.Linq.Expressions\System\Linq\Expressions\Interpreter\LightLambda.cs:line 274
   at System.Func`2.Invoke(T arg)
   at System.Func`2.InvokeObjectArrayThunk(T arg)
   at System.Func`2.Invoke(T arg)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass20_0`1.<CompileQueryCore>b__0(QueryContext qc)
   at System.Func`2.Invoke(T arg)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at System.Linq.Queryable.FirstOrDefault[TSource](IQueryable`1 source, Expression`1 predicate)
   at TVPLAN.Models.Data.SeriesStore.GetSeriesTitle(SeriesID ID) in C:\Users\Willi\Documents\Programming\Source Control (VSO)\TVPlan\Future-Branch\TVPLAN\TVPLAN-Shared\Models\Data\SeriesStore.cs:line 91
   at TVPLAN.ViewModels.Library.LibraryItem..ctor(SeriesID ID, CoreDispatcher dispatcher) in C:\Users\Willi\Documents\Programming\Source Control (VSO)\TVPlan\Future-Branch\TVPLAN\TVPLAN\ViewModels\Library\LibraryItem.cs:line 29
   at TVPLAN.ViewModels.Library.LibraryModel..ctor(CoreDispatcher LibraryPageDispatch) in C:\Users\Willi\Documents\Programming\Source Control (VSO)\TVPlan\Future-Branch\TVPLAN\TVPLAN\ViewModels\Library\LibraryModel.cs:line 37
   at TVPLAN.Views.Library..ctor() in C:\Users\Willi\Documents\Programming\Source Control (VSO)\TVPlan\Future-Branch\TVPLAN\TVPLAN\Views\Library.xaml.cs:line 44
   at TVPLAN.TVPLAN_XamlTypeInfo.XamlTypeInfoProvider.Activate_79_Library() in C:\Users\Willi\Documents\Programming\Source Control (VSO)\TVPlan\Future-Branch\TVPLAN\TVPLAN\obj\x86\Release\XamlTypeInfo.g.cs:line 496
   at WinRTXamlToolkit.WinRTXamlToolkit_XamlTypeInfo.XamlUserType.ActivateInstance()
   at __Interop.ReverseComStubs.Stub_9(Object __this, Void** unsafe___value__retval, IntPtr __methodPtr) in C:\Users\Willi\Documents\Programming\Source Control (VSO)\TVPlan\Future-Branch\TVPLAN\TVPLAN\obj\x86\Release\ilc\intermediate\TVPLAN.McgInterop\SharedStubs.g.cs:line 28292
```

### Steps to reproduce
If you require the full solution than I can include emails, as it is a Private VSOnline Repo.

```c#
    public class SeriesStore : DbContext
    {
        ...
        public static string GetSeriesTitle(SeriesID ID)
        {
            using (var cache = new SeriesStore())
            {
                return cache.TVDBSeries.FirstOrDefault(item => item.Id == ID.TVDB)?.Title ?? "";
            }
        }
        ....
     }
```

I have tried setting Optimise Code to off in Build Properties on both my Main App Project and my Shared Project, the Shared Project nothing happens, while on Main App Project the app crashes instantly.

I am aware of #6253 and I have tried adding to both Default.rd & TVPLAN_Shared.rd:
```XAML
    <Assembly Name="Microsoft.EntityFrameworkCore.Sqlite" DoNotOptimize="true" DoNotInline="true" />
    <Assembly Name="Microsoft.EntityFrameworkCore.Tools" DoNotOptimize="true" DoNotInline="true" />
```
This doesn't appear to do anything.

Adding the following:
```XAML    
<Type Name="TVPLAN.Models.Data.SeriesStore" DoNotOptimize="true" DoNotInline="true" />
```
Causes the exception to change to `ArgumentNull_Generic`.

```XAML    
<Type Name="TheTVDBSharp.Models.Series" DoNotOptimize="true" DoNotInline="true" />
```
Causes the exception to change to something along the lines of `Parameter{0} ...`, I didn't capture it, and I won't endure another 5 minutes to receive the result unless you ask me to, also now I can't seem to replicate it.

Adding `DoNotOptimize="true" DoNotInline="true"` to the assembly or to `System.Nullable{T}`, with `<UseDotNetNativeSharedAssemblyFrameworkPackage>false</UseDotNetNativeSharedAssemblyFrameworkPackage>`
Doesn't change the result, and crashes VS 2017 when showing the same result! 

**All of these tests were applied to both the Main App Project and the Shared Library Project.**

People have said that Upgrading to 5.3.0 of `Microsoft.NETCore.UniversalWindowsPlatform ` would help, but this predicament started on 5.3.2. I have combed through all of the similar issues in this Repo, to no avail. :(

If I get this method to return an empty string, the app runs, but it vomits caught Exceptions in my In-App Exception Dialog, where another Database is causing issues with my Syncing Service :(

### Further technical details
Microsoft.NETCore.UniversalWindowsPlatform Version: 5.3.2
EF Core Packages Installed (In Both Projects):
**Microsoft.EntityFrameworkCore.SQLite V1.1.1
Microsoft.EntityFrameworkCore.Tools V1.1.0**
Database Provider: Microsoft.EntityFrameworkCore.SQLite
Operating system: Windows 10 Pro Creators Update
IDE: Visual Studio 2017 (15.1 26403.3)

**Comment by ajcvickers** [comment]
@WilliamABradley Thanks for reporting this. Can you give us access to the full solution. I am "avickers" at this place I work.

**Comment by WilliamABradley** [comment]
@ajcvickers Is there an email so I can add you to my VSO Readers group, or is there a better way to do this?

**Comment by WilliamABradley** [comment]
@ajcvickers  I sent you the Solution via OneDrive Email, Shared to your email.

**Comment by ajcvickers** [comment]
Note for triage: I have the repro files, so we can assign this now.

**Comment by WilliamABradley** [comment]
Are there any updates as to what can be done to fix this?

**Comment by ajcvickers** [comment]
@WilliamABradley It takes more time to reproduce and debug ,NET Native issues. We currently don't have any resources free to do so. We are working towards getting someone available to investigate.

@MattWhilden If someone from .NET Native is available sooner to investigate this, then I can send over the customer's repro project.

**Comment by WilliamABradley** [comment]
Ok, thanks

**Comment by WilliamABradley** [comment]
@ajcvickers  @MattWhilden Have you managed to get someone to start investigating the issue? I'm looking to release the update from half way through next month, but perhaps it could last until the end of the month.
Thanks,
William.

**Comment by ajcvickers** [comment]
@WilliamABradley I'll try to do the investigation this week.

**Comment by WilliamABradley** [comment]
@ajcvickers Awesome,
Thanks.

**Comment by ajcvickers** [comment]
@WilliamABradley I was able to run the application, but I am not able to make it fail with the exception you are seeing. What I am seeing is that the DbSet properties on the context are null, so I get an ArgumentNull exception instead. But as of yet I can't figure out why the DbSet properties are null--this should never happen. I will do some more investigation tomorrow, but if you have any additional pointers on reproducing the issue it would be great to know.

**Comment by ajcvickers** [comment]
@WilliamABradley I was able to reproduce the issue and do some investigation. Things I tried:
* Using EF 1.1.2, which has some Reflection fixes--didn't help
* Using Native 5.3.3--didn't help
* Using LINQ but forcing client eval at various levels--didn't help. (A simple ToLIst on the entity set repros the problem.)
* Even FromSql (`cache.TVDBSeries.FromSql("select * from TVDBSeries").ToList();`) has the issue

I tried doing some debugging but wasn't able to pin down the root cause. Marking this for re-triage to figure out the best way forward. Some points to consider:
* It looks like it could be a Reflection issue that we have missed
  * The model is using unsigned types (mentioning it because there is not much coverage and we have had issues in this area, but don't see anything that indicates it is the problem here.)
  * The types that are getting mixed up are nullable enums--could be something in the unwrapping/unerlying type checking that is not playing well with .NET Native and/or Reflection comparisons
* It could just be a .NET Native bug

**Comment by ajcvickers** [comment]
@WilliamABradley Had an idea while napping and just tried it. I replaced all the nullable enums in the model with `int?` properties. Success! So I hope this means two things:
* You can unblock yourself by making this change in your model--you'll have to add a few casts where you use the properties, but from what I could see it didn't seem too bad.
* We now have a more concrete lead as to what is not working in EF/Native--nullable enum properties.

**Comment by WilliamABradley** [comment]
@ajcvickers That's awesome! I'll let you know if I run into any more issues. Would be another solution be two replace the nullable Enums with Enums that have a default of say "Unknown"?

**Comment by ajcvickers** [comment]
@WilliamABradley Possibly, but I didn't try that.

**Comment by WilliamABradley** [comment]
@ajcvickers Do I need to change the properties in the actual model, or does changing the Types in the migrations work?

**Comment by WilliamABradley** [comment]
@ajcvickers FYI I got it working by switching my Nullable Enums with Nullable Ints :) 
Thanks for the help.

**Comment by AndriySvyryd** [comment]
Verified that this is fixed in the prerelease builds of EF 2.0.1 and UWP 6.0.0
